units/event_log_tests.cc

```diff
@@ -0,0 +1,183 @@
+      : name(_name), log_level(_log_level) {}
+
+  // Each implementation of an Event contains attribute(s) and
+  // overrides this function to return them.
+  virtual const std::vector<Attribute *> &GetAttributes() = 0;
+  virtual size_t GetNumAttributes() = 0;
+
+  const char *GetEventName() { return name; }
+  LogLevel GetLogLevel() { return log_level; }
+};
+
+// An event for the CreateShaderModule function.
+// Example:
+//    const int64_t create_time_ns = create_end - create_start;
+//    Int64Attr timestamp("timestamp", 1234);
+//    Int64Attr shader_hash("hash", res.shader_hash);
+//    std::vector<Attribute *> attributes = {&shader_hash, &timestamp};
+//    CreateShaderModuleEvent("create_shader_module_ns", attributes,
+//    LogLevel::High);
+struct CreateShaderModuleEvent : Event {
+  std::vector<Attribute *> attributes;
+
+  CreateShaderModuleEvent(const char *_name,
+                          std::vector<Attribute *> &_attributes,
+                          LogLevel _log_level)
+      : Event(_name, _log_level), attributes(_attributes) {}
+  const std::vector<Attribute *> &GetAttributes() override {
+    return attributes;
+  }
+  size_t GetNumAttributes() override { return attributes.size(); }
+};
+
+// Takes an `Event` instance as an input and generates a csv string containing
+// `event`'s name and attribute values.
+std::string EventToCSVString(Event &event) {
+  const std::vector<Attribute *> &attributes = event.GetAttributes();
+
+  std::string csv_str = event.GetEventName();
+  csv_str += ",";
+  size_t e;
+  for (size_t i = 0, e = attributes.size(); i != e; ++i) {
+    switch (attributes[i]->value_type) {
+      case AttributeType::Int64: {
+        csv_str +=
+            ValueToCSVString(attributes[i]->cast<Int64Attr>()->GetValue());
+        break;
+      }
+      case AttributeType::String: {
+        csv_str +=
+            ValueToCSVString(attributes[i]->cast<StringAttr>()->GetValue());
+        break;
+      }
+      case AttributeType::VectorInt64: {
+        csv_str += ValueToCSVString(
+            attributes[i]->cast<VectorInt64Attr>()->GetValue());
+        break;
+      }
+    }
+    if (i + 1 != e) csv_str += ",";
+  }
+  return csv_str;
+}
+
+TEST(Event, AttributeCreation) {
+  Int64Attr timestamp("timestamp", 1601314733266018318);
+  StringAttr state("state", "1");
+  VectorInt64Attr pipeline("pipeline",
+                           {0x67d6fd0aaa78a6d8, 0x67d390249c2f20ce});
+  SUCCEED();
+}
+
+TEST(Event, EventCreation) {
+  Int64Attr timestamp("timestamp", 1601314733266018318);
+  StringAttr state("state", "1");
+  VectorInt64Attr pipeline("pipeline",
+                           {0x67d6fd0aaa78a6d8, 0x67d390249c2f20ce});
+  std::vector<Attribute *> compile_attrs = {&timestamp, &state, &pipeline};
+  CreateShaderModuleEvent compile_event("compile_time", compile_attrs,
+                                        LogLevel::Low);
+  EXPECT_EQ(compile_event.GetNumAttributes(), 3);
```

Comment:
Use `ASSERT_EQ` instead. The difference is that `ASSERT_*` checks abort execution of the current test on failure, while `CHECK_*` continue executing. When the size is not as expected, we shouldn't try to access vector elements that don't exist.